klib/tun.c

```diff
@@ -169,130 +244,202 @@ closure_function(1, 6, sysreturn, tun_write,
+{
+    if (!tun_cfg)
+        return;
+    tuple cfg = kfuncs.get(tun_cfg, sym_this(name));
+    if (!cfg || !is_tuple(cfg))
+        return;
+    buffer ipb = kfuncs.get(cfg, sym_intern(ipaddress, kfuncs.intern));
+    if (ipb) {
+        char *ip = buffer_to_cstring(ipb);
+        if (!kfuncs.ip4addr_aton(ip, ipaddr)) {
+            kfuncs.rprintf("tun: invalid ipaddress %s\n", ip);
+        }
+    }
+    buffer nmb = kfuncs.get(cfg, sym_intern(netmask, kfuncs.intern));
+    if (nmb) {
+        char *nm = buffer_to_cstring(nmb);
+        if (!kfuncs.ip4addr_aton(nm, netmask) || !kfuncs.ip4_addr_netmask_valid(netmask->addr)) {
+            kfuncs.rprintf("tun: invalid netmask %s\n", nm);
+        }
+    }
+
+    *bringup = kfuncs.get(cfg, sym_intern(up, kfuncs.intern)) != 0;
+}
+
 closure_function(1, 2, sysreturn, tun_ioctl,
-                 tun, tun,
+                 tun_file, tf,
                  unsigned long, request, vlist, ap)
 {
-    tun tun = bound(tun);
+    tun_file tf = bound(tf);
+    tun tun = tf->tun;
     switch (request) {
     case TUNSETIFF: {
         struct ifreq *ifreq = varg(ap, struct ifreq *);
         if (!validate_user_memory(ifreq, sizeof(struct ifreq), true))
             return -EFAULT;
-        if (tun->netif || kfuncs.netif_find(ifreq->ifr_name))
+        if (tun)
             return -EINVAL;
         switch (ifreq->ifr.ifr_flags & TUN_TYPE_MASK) {
         case IFF_TUN:
             break;
         default:
             return -EINVAL;
         }
-        if ((ifreq->ifr.ifr_flags & ~TUN_TYPE_MASK) & ~IFF_NO_PI)
+        if ((ifreq->ifr.ifr_flags & ~TUN_TYPE_MASK) & ~(IFF_NO_PI | IFF_MULTI_QUEUE))
             return -EINVAL;
-        tun->flags = ifreq->ifr.ifr_flags;
-        tun->netif = allocate(tun_heap, sizeof(struct netif));
-        if (tun->netif == INVALID_ADDRESS)
-            return -ENOMEM;
-        if (ifreq->ifr_name[0] && ifreq->ifr_name[1]) {
-            tun->netif->name[0] = ifreq->ifr_name[0];
-            tun->netif->name[1] = ifreq->ifr_name[1];
-        } else {    /* assign a default name */
-            tun->netif->name[0] = 't';
-            tun->netif->name[1] = 'u';
+        struct netif *netif = kfuncs.netif_find(ifreq->ifr_name);
+        if (netif) {
+            if (netif->output != tun_if_output)
+                return -EINVAL;
+            tun = netif->state;
+        } else {
+            tun = allocate(tun_heap, sizeof(struct tun));
+            if (tun == INVALID_ADDRESS)
+                return -ENOMEM;
+            spin_lock_init(&tun->lock);
+            tun->flags = ifreq->ifr.ifr_flags;
+            if (ifreq->ifr_name[0] && ifreq->ifr_name[1]) {
+                tun->netif.name[0] = ifreq->ifr_name[0];
+                tun->netif.name[1] = ifreq->ifr_name[1];
+            } else {    /* assign a default name */
+                tun->netif.name[0] = 't';
+                tun->netif.name[1] = 'u';
+            }
+            tun->netif.state = tun;
```

Comment:
`tun` is passed as `state` argument to netif_add(), so there is no need to assign it here too. (again, my fault)

Reply:
Thanks!